Feature 5930 unsetting the child rule

ThePhisch opened the pull request:
* putting in `child_rule_id` as `None` in `update_rule` will unset a previously set child rule.
* wrote tests (`test_bin_rucio` and `test_rule`)

* also included: a small edit solving the `execute` function that is mainly used in `test_bin_rucio`. originally, this function had two different return signatures, depending on the value of a keyword-argument `blocking`, which defaulted to `True`. This duality caused the linter/type checker to raise a warning each time the execute-function was called and unpacked into the three elements of the tuple it would normally return (ie when blocking is True). since the function is called frequently during the testing process, a file such as `test_bin_rucio` might have >200 warnings. the solution was to keep the `execute`-function only in the default form (returning a tuple of three elements) and (since the other form was only used once in the entire codebase) replace the function call with `blocking=False` with the code block within the `execute` function that would have been called otherwise.

Commit:
Common: Refactor execute to single return signature better typehints; Fix #5942

The function rucio/common/utils.py::execute is of great importance for
the testing workflows, since it allows for shell commands to be run and
their outputs to be used. The issue was that the function had two return
signatures, depending on the value of a keyword argument
(blocking=True). The different return signatures cause linting errors.

The function was refactored to have a single return signature, removing
the keyword argument and only treating the case where blocking=True. The
one situation in which the function 'execute' is called with
blocking=False has been handled by substituting the body of the
'execute' function into what was the function call.

dchristidis reviewed (commented):
I’m guessing that one of the two commits shouldn’t be part of this PR?

ThePhisch commented:
> 

no both are intended to be in this PR. I considered not opening an issue for the first one but I decided to do so anyway because I changed the functionality of a frequently-used function in order to provide better type hints.

dchristidis reviewed (commented):
I think I prefer the approach used in `move-rule`, where `None` is not case-sensitive and is not passed as a string to the API. See commit d440f9353a.

ThePhisch commented:
I reworked my commit to mirror d440f93 as requested

dchristidis commented on lib/rucio/api/rule.py:
```diff
@@ -228,7 +230,7 @@ def delete_replication_rule(rule_id, purge_replicas, issuer, vo='def', session=N
 
 @transactional_session
-def update_replication_rule(rule_id, options, issuer, vo='def', session=None):
+def update_replication_rule(rule_id: str, options: Dict[str, Any], issuer: str, vo='def', session=None):
```
Why avoid type hints for `vo` and `session`?

ThePhisch replied:
in the case of vo, mypy can infer the type from the default value given. i consider it to be redundant information to add a `str` tye-hint.
in the case of `session`, Radu is working on this already https://github.com/rcarpa/rucio/commit/94bcc0940451545628e0e53f5feb324d7911c665

dchristidis replied:
For type hints in parameters, I find that ‘explicit is better than implicit’, as instructed by _The Zen of Python_. But I won’t push you.

ThePhisch replied:
you are right. consider it done

dchristidis commented on lib/rucio/client/ruleclient.py:
```diff
@@ -107,7 +108,7 @@ def get_replication_rule(self, rule_id):
             raise exc_cls(exc_msg)
 
-    def update_replication_rule(self, rule_id, options):
+    def update_replication_rule(self, rule_id, options: Dict[str, Any]):
```
Ditto for `rule_id`.

ThePhisch replied:
will fix this.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1180,14 +1180,18 @@ def update_rule(rule_id, options, session=None):
     """
 
-    valid_options = ['comment', 'locked', 'lifetime', 'account', 'state', 'activity', 'source_replica_expression', 'cancel_requests', 'priority', 'child_rule_id', 'eol_at', 'meta', 'purge_replicas', 'boost_rule']
+    valid_options = ['comment', 'locked', 'lifetime', 'account', 'state',
+                     'activity', 'source_replica_expression', 'cancel_requests',
+                     'priority', 'child_rule_id', 'eol_at', 'meta',
+                     'purge_replicas', 'boost_rule']
```
Isn’t this purely a cosmetic change? I find that bundling it in a commit with functional changes can be confusing.

ThePhisch replied:
it is a purely cosmetic change, i can roll it back if you feel strongly about that.

dchristidis replied:
No, I do not object, but I would advise having a separate commit for this. Same for the addition of type hints (especially in unrelated code).

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1285,11 +1289,24 @@ def update_rule(rule_id, options, session=None):
-                    raise InputValidationError('Parent and child rule must be set on the same dataset.')
-                if child_rule.state != RuleState.OK:
-                    rule.child_rule_id = options[key]
+                child_id: Union[str, None] = options[key]
```
This is what `Optional` is for.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1285,11 +1289,24 @@ def update_rule(rule_id, options, session=None):
+                    # dissolve exisiting relationship
+                    if not rule.child_rule_id:
+                        raise InputValidationError('Cannot detach child when '
+                                                   'no such relationship exists')
```
I’m starting to feel strongly against breaking string with errors message for the sake of maintaining a maximum line length. It hinders searching for them with `grep`. I’m not convinced that it is worthwhile.

dchristidis commented on lib/rucio/web/rest/flaskapi/v1/rules.py:
```diff
@@ -117,6 +118,9 @@ def put(self, rule_id):
                         type: string
+                      unset_child:
+                        description: Whether to unset the child rule.
+                        type: boolean
```
This appears to be a leftover from your previous work.

ThePhisch replied:
true, will fix this

dchristidis commented on lib/rucio/web/rest/flaskapi/v1/rules.py:
```diff
@@ -150,14 +154,16 @@ def put(self, rule_id):
         parameters = json_parameters()
-        options = param_get(parameters, 'options')
+        options: Dict[str, Any] = param_get(parameters, 'options', default={})
+        print(options)
```
Ditto.

ThePhisch replied:
i added type hints to the options dict and a default value, i dont think that that has to go. on the other hand, i will remove the `print` statement.

Commit:
Core: Unset Child Rule is now possible in Update_Rule; Fix #5930

Expanded the update_rule function. If 'None' is given as the new name of
the child rule, the currently attached child rule is detached. This
functionality comes with tests added to test_bin_rucio and test_rule.
The CLI parser in rucio has also been updated.

Commit:
Decorative Changes

Replace issuer with issuer_ia in rucio/api/rule.py because 'issuer' is
an argument of type str and overwriting it with an InteralAccount causes
linting errors.

Type hints added, the types are inferred from bin/rucio and the
specifications given in rucio/web/rest/flaskapi/v1/rules.py::AllRule

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1187,7 +1187,8 @@ def update_rule(rule_id, options, session=None):
     try:
-        rule = session.query(models.ReplicationRule).filter_by(id=rule_id).one()
+        rule: models.ReplicationRule = session.query(models.ReplicationRule).\
+            filter_by(id=rule_id).one()
```
Based on what we discussed, this shouldn’t be part of this commit.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1285,14 +1286,24 @@ def update_rule(rule_id, options, session=None):
-                    rule.child_rule_id = options[key]
+                child_id: Optional[str] = options[key]
+                if child_id is None:
+                    # dissolve exisiting relationship
```
I find this comment redundant.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1285,14 +1286,24 @@ def update_rule(rule_id, options, session=None):
+                    rule.expires_at = None  # type: ignore
+                else:
+                    child_rule = (session.query(models.ReplicationRule).
+                                  filter_by(id=child_id).one())
```
This would look better like so:
```python
                    child_rule = (session.query(models.ReplicationRule)
                                         .filter_by(id=child_id)
                                         .one())
```

dchristidis replied:
Bonus points if you use the SQLAlchemy 2.0 syntax instead! Refer to #5247 if you’re willing, but note that it would probably have to go into its own commit.

dchristidis reviewed (changes requested):
The commit message of e90284f6a7bcd531118a99d1a2d4f7b8115d7ecc is no longer valid and should be rewritten. Please provide some context as to *why* you’re doing what you’re doing.

When adding type hints, please try to address all parameters and the return type. It’s good that you’re doing it for `rule_id`, but I think it’s better to do the entire function (especially since some appear to be trivial to complete.

dchristidis commented on lib/rucio/web/rest/flaskapi/v1/rules.py:
```diff
@@ -150,14 +150,15 @@ def put(self, rule_id):
        """
         parameters = json_parameters()
-        options = param_get(parameters, 'options')
+        options = param_get(parameters, 'options', default={})
```
Doesn’t this require some special mention?

ThePhisch replied:
i decided to take it out. I want the error for a missing parameter to be raised as early as possible, not later.

dchristidis replied:
Did you actually take it out?

dchristidis commented on lib/rucio/api/rule.py:
```diff
@@ -246,11 +248,11 @@ def update_replication_rule(rule_id, options, issuer, vo='def', session=None):
-        issuer = InternalAccount(issuer, vo=vo)
+        issuer_ia = InternalAccount(issuer, vo=vo)
         if options['approve']:
-            rule.approve_rule(rule_id=rule_id, approver=issuer, session=session)
+            rule.approve_rule(rule_id=rule_id, approver=issuer_ia, session=session)
         else:
-            rule.deny_rule(rule_id=rule_id, approver=issuer, reason=options.get('comment', None), session=session)
+            rule.deny_rule(rule_id=rule_id, approver=issuer_ia, reason=options.get('comment', None), session=session)
```
Why is this better?

ThePhisch replied:
because issuer is an argument of type string. Overwriting it with an InternalAccount causes linting errors.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+                    models.ReplicaLock.state == LockState.REPLICATING
+                )
+                replicalocks = session.execute(query).all()
+                for lock, in replicalocks:
```
Use `.scalars().all()` (and maybe merge into one line?)

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+                locks = session.execute(query).all()
                 counter_rses = {}
-                for lock in locks:
+                for lock, in locks:
```
Ditto.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+                            models.ReplicaLock.state == LockState.REPLICATING
+                        )
+                        locks2 = session.execute(query).all()
+                        for lock2, in locks2:
```
Ditto.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
-                            models.RSEFileAssociation.rse_id == lock.rse_id).one()
+                            models.RSEFileAssociation.rse_id == lock.rse_id
+                        )
+                        replica, = session.execute(query).one()
```
Ditto.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1190,7 +1191,13 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+        ).where(
+            models.ReplicationRule.id == rule_id
+        )
+        rule, = session.execute(query).one()
```
Use `scalar_one()` instead.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+
+                    locks = session.execute(query).all()
+                    for lock, in locks:
+                        lock.update({'account': options['account']})
```
Not sure I like this.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1208,62 +1215,162 @@ def update_rule(rule_id: str, options: Dict[str, Any], session=None) -> None:
+                        models.ReplicaLock.state == LockState.REPLICATING
+                    )
+                    locks = session.execute(query).all()
+                    for lock, in locks:
```
Ditto.

dchristidis reviewed (changes requested):
Your last commit combines functional and aesthetic changes, which makes it more difficult to review than it should be.

> The commit message of https://github.com/rucio/rucio/commit/e90284f6a7bcd531118a99d1a2d4f7b8115d7ecc is no longer valid and should be rewritten. Please provide some context as to why you’re doing what you’re doing.

This remains to be addressed, but for commit 4caa9dbe79746ac541ad6f4b4dbd3327834a94b1.

dchristidis commented on lib/rucio/tests/test_rule.py:
```diff
@@ -952,6 +952,39 @@ def test_update_rule_child_rule(self, mock_scope, did_factory, jdoe_account):
+
+        # test detaching when there is no more attached child
+        with pytest.raises(InputValidationError):
+            update_rule(parent_id, {'unset_child': True})
```
No bueno!

dchristidis commented on lib/rucio/client/ruleclient.py:
```diff
@@ -192,7 +218,9 @@ def deny_replication_rule(self, rule_id):
 
-    def list_replication_rule_full_history(self, scope, name):
+    def list_replication_rule_full_history(
+            self, scope: Union[str, bytes], name: Union[str, bytes]
```
Why `bytes`?

ThePhisch replied:
scope and name are used as arguments for the `quote_plus` function, which takes `str` or `bytes`. This was the only upper bound to the types applicable to scope and name I could find.

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1231,15 +1251,26 @@ def update_rule(rule_id: str, options: Dict[str, Any], *, session: "Session") ->
+                    ).values(
+                        account=options['account']
+                    )
+
```
Is this query executed?

ThePhisch replied:
no

ThePhisch replied:
it is now

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1251,23 +1282,61 @@ def update_rule(rule_id: str, options: Dict[str, Any], *, session: "Session") ->
             elif key == 'state':
                 if options.get('cancel_requests', False):
                     rule_ids_to_stuck = set()
-                    for lock in session.query(models.ReplicaLock).filter_by(rule_id=rule.id, state=LockState.REPLICATING).all():
+                    query = select(
+                        models.ReplicaLock
+                    ).where(
+                        models.ReplicaLock.rule_id == rule.id,
+                        models.ReplicaLock.state == LockState.REPLICATING
+                    )
+                    for lock in session.execute(query).scalars().all():
                         # Set locks to stuck:
-                        for lock2 in session.query(models.ReplicaLock).filter_by(scope=lock.scope, name=lock.name, rse_id=lock.rse_id, state=LockState.REPLICATING).all():
+                        query = select(
+                            models.ReplicaLock
+                        ).where(
+                            models.ReplicaLock.scope == lock.scope,
+                            models.ReplicaLock.name == lock.name,
+                            models.ReplicaLock.rse_id == lock.rse_id,
+                            models.ReplicaLock.state == LockState.REPLICATING
+                        )
+                        for lock2 in session.execute(query).scalars().all():
                             lock2.state = LockState.STUCK
                             rule_ids_to_stuck.add(lock2.rule_id)
-                        transfers_to_cancel = request_core.cancel_request_did(scope=lock.scope, name=lock.name, dest_rse_id=lock.rse_id, session=session)
+                        transfers_to_cancel = request_core.cancel_request_did(
+                            scope=lock.scope,
+                            name=lock.name,
+                            dest_rse_id=lock.rse_id,
+                            session=session
+                        )
```
Isn’t this purely decorative?

ThePhisch replied:
yup

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1251,23 +1282,61 @@ def update_rule(rule_id: str, options: Dict[str, Any], *, session: "Session") ->
+                            models.DatasetLock.rule_id == rid
+                        )
+                        dataset_lock = session.execute(query).scalar_one()
+                        dataset_lock.update({'state': LockState.STUCK})
```
Why SELECT then UPDATE?

ThePhisch replied:
to make sure that exactly one is updated. but this assumption is false, since multiple ReplicationRules/DatasetLocks can be filtered and updated

dchristidis commented on lib/rucio/core/rule.py:
```diff
@@ -1276,8 +1345,22 @@ def update_rule(rule_id: str, options: Dict[str, Any], *, session: "Session") ->
+                            models.DatasetLock.rule_id == rule.id
+                        )
+                        dataset_lock = session.execute(query).scalar_one()
+                        dataset_lock.update({'state': LockState.STUCK})
```
Ditto.

dchristidis reviewed (changes requested):
Almost there!

Commit:
Core: All core.rule.py::update_rule follows SQLAlchemy2 Syntax; Fix #5247

Commit:
Decorative Changes

dchristidis reviewed (dismissed):
No further comments. Thanks for bearing with me.

Unfortunately, a rebase on the latest master is necessary.
